libs/wingc/src/lsp/document_symbols.rs

```diff
@@ -0,0 +1,126 @@
+use crate::ast::{Scope, *};
+use crate::lsp::ast_traversal::{
+	get_expressions_from_statement, get_statements_from_scope_no_context, get_symbols_from_expression,
+	get_symbols_from_statement, TreeLocationContext,
+};
+use crate::lsp::sync::FILES;
+use crate::wasm_util::{ptr_to_string, string_to_combined_ptr};
+use lsp_types::DocumentSymbol;
+use lsp_types::SymbolKind;
+
+#[no_mangle]
+pub unsafe extern "C" fn wingc_on_document_symbol(ptr: u32, len: u32) -> u64 {
+	let parse_string = ptr_to_string(ptr, len);
+	if let Ok(parsed) = serde_json::from_str(&parse_string) {
+		if let Some(token_result) = on_document_symbols(parsed) {
+			let result = serde_json::to_string(&token_result).unwrap();
+
+			string_to_combined_ptr(result)
+		} else {
+			0
+		}
+	} else {
+		eprintln!("Failed to parse 'onHover' text document: {}", parse_string);
+		0
+	}
+}
+
+pub fn on_document_symbols<'a>(params: lsp_types::DocumentSymbolParams) -> Option<Vec<DocumentSymbol>> {
+	FILES.with(|files| {
+		let files = files.borrow();
+		let files = files.read();
+		let files = files.unwrap();
+		let parse_result = files.get(&params.text_document.uri);
+		let parse_result = parse_result.unwrap();
+		let scope = &parse_result.scope;
+
+		let symbols = create_symbols(scope);
+
+		if symbols.is_empty() {
+			None
+		} else {
+			Some(symbols)
+		}
+	})
+}
+
+fn create_symbols<'a>(scope: &'a Scope) -> Vec<DocumentSymbol> {
+	let mut document_symbols: Vec<DocumentSymbol> = vec![];
+	let mut raw_symbols: Vec<(TreeLocationContext<'a>, &'a Symbol)> = vec![];
+
+	let all_statements = get_statements_from_scope_no_context(scope);
+	for context in all_statements {
+		raw_symbols.extend(get_symbols_from_statement(context));
+
+		let all_expressions = get_expressions_from_statement(context);
+		for expression_context in all_expressions {
+			raw_symbols.extend(get_symbols_from_expression(expression_context));
+		}
+	}
+
+	for symbol_lookup in raw_symbols {
+		let statement = symbol_lookup.0.statement;
+		let symbol = symbol_lookup.1;
+		let symbol_kind = match &statement.kind {
+			StmtKind::Bring { .. } => SymbolKind::NAMESPACE,
+			StmtKind::VariableDef { var_name, .. } => {
+				if var_name.name == symbol.name {
+					SymbolKind::VARIABLE
+				} else {
+					continue;
+				}
+			}
+			StmtKind::ForLoop { iterator, .. } => {
+				if iterator.name == symbol.name {
+					SymbolKind::VARIABLE
+				} else {
+					continue;
+				}
+			}
+			StmtKind::Class(c) => {
+				if c.name.name == symbol.name {
+					SymbolKind::CLASS
+				} else {
+					// eventually we should handle the entire class symbol hierarchy
```

Comment:
What's an example of a case the code won't handle? (just curious)

Reply:
A big part of the DocumentSymbol thing is that it can represent hierarchy. e.g.

```
variable
closure
	class
		method
		variable
```

This whole file is just making all symbols flat, as if they were all defined at the root. It's not useless right now, but it is inaccurate.